Fixes issue related torch.trapezoid broadcasting behavior and documentation

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64054

Fixes #63608


cc @mruberry @rgommers @heitorschueroff

Differential Revision: [D30617078](https://our.internmc.facebook.com/intern/diff/D30617078)

NivekT committed:
Fixes #63608 issue related torch.trapezoid broadcasting behavior and documentation

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/Integration.cpp:
```diff
@@ -53,6 +53,14 @@ Tensor do_cumulative_trapezoid(const Tensor& y, double dx, int64_t dim) {
     return (dx /2. * (left + right)).cumsum(dim);
 }
 
+DimVector add_padding(IntArrayRef curr_shape, int64_t target_size) {
```
Add a comment describing what this function does and how it's intended to be used

NivekT replied:
Done. I also changed the names of the function and its variables for more clarity.

mruberry commented on aten/src/ATen/native/Integration.cpp:
```diff
@@ -53,6 +53,14 @@ Tensor do_cumulative_trapezoid(const Tensor& y, double dx, int64_t dim) {
 
+DimVector add_padding(IntArrayRef curr_shape, int64_t target_size) {
+    DimVector new_sizes(target_size, 1);
+    for (int i = 0; i < curr_shape.size(); i++) {
```
type `i` like `curr_shape.size()` (I prefer using decltype but whatever)

NivekT replied:
If I am understanding correctly, I should use `decltype(curr_shape.size()) i = 0` in place of `int i = 0`?

If so, done.

mruberry replied:
That's one way to do it. You can also do `auto i = decltype(curr_shape.size()){0}`, which is a common pattern, or find what the size type of curr_shape is and just type `i` directly as that

mruberry commented on test/test_binary_ufuncs.py:
```diff
@@ -2664,6 +2664,11 @@ def test_x(sizes, dim, x, device):
         test_x((1, 10), 0, [1.0], device)
         test_x((0, 2), 0, [], device)
         test_x((0, 2), 1, [1.0, 2.0], device)
+        test_x((2, 3, 4), -1, [1.0, 2.0, 3.0, 4.0], device)
```
Would these tests have failed previously?

NivekT replied:
Yes, those are similar to the cases that are failing in issue #63608

mruberry commented on torch/_torch_docs.py:
```diff
@@ -10865,11 +10865,11 @@ def merge_dicts(*dicts):
+and :attr:`y`, the function computes the difference between consecutive elements along
+dimension :attr:`dim`. This effectively creates two tensors, `x_diff` and `y_diff`, that have
+the same shape as the original tensors except their lengths along the dimension :attr:`dim` is reduced by 1.
+After that, those two tensors are broadcasted together to compute final output as part of the trapezoidal rule.
```
"broadcast" is the past tense of "broadcast," not "broadcasted" (analogous to "cast" being the past tense of "cast," and not "casted")

NivekT replied:
Noted :)

mruberry commented on torch/_torch_docs.py:
```diff
@@ -10865,11 +10865,11 @@ def merge_dicts(*dicts):
-then that dimension defines the spacing for each computation.
-If :attr:`x` is also specified and is not one-dimensional, then it is broadcast to
-the shape of :attr:`y` and the corresponding sizes are used for each computation.
+The broadcasting behavior of this function is as follows. First, for both :attr:`x`
```
This first sentence and the following "First," seem like a lot of introduction because there's no "Second..." Maybe just talk about how the computation occurs when x and y are the same size vs. different sizes?

NivekT replied:
I slightly modified the paragraph. Let me know if it is still unclear.

mruberry reviewed (commented):
This is really cool, @NivekT, thanks for the follow-up.

I made a few comments inline for your review, nothing big. Would you also update the PR description to clarify what was happening and how this fix addresses the issue?

NivekT committed:
Update on "Fixes issue related torch.trapezoid broadcasting behavior and documentation"


Fixes #63608


cc @mruberry @rgommers @heitorschueroff

Differential Revision: [D30617078](https://our.internmc.facebook.com/intern/diff/D30617078)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30617078).

mruberry commented on aten/src/ATen/native/Integration.cpp:
```diff
@@ -52,7 +52,21 @@ Tensor do_cumulative_trapezoid(const Tensor& y, double dx, int64_t dim) {
     return (dx /2. * (left + right)).cumsum(dim);
 }
-
+// Given the current shape of a Tensor and a target number of dimensions,
```
Why not just use expand for this?

https://pytorch.org/docs/master/generated/torch.Tensor.expand.html#torch.Tensor.expand

NivekT replied:
The behavior that we want (and `numpy` has) is for `x` and `y` tensors to broadcast together.

If we do `x.expand(y.sizes())`, we would be broadcasting `x` into the shape of `y`. Therefore, in this part, I create a custom function to add padding of `1`s to the shape of `x` when it is has a smaller dimension than `y`.

mruberry replied:
They could be expanded to the common size, though. There's also broadcast_tensors():

https://pytorch.org/docs/master/generated/torch.broadcast_tensors.html?highlight=broadcast_tensors#torch.broadcast_tensors

It'd be weird if we didn't have an existing function for this

NivekT replied:
Oh one more thing, `numpy` also broadcast after taking the difference of each tensor. If we broadcast here, it would be before we take the difference of the tensors.

mruberry replied:
You mean this is still divergent with NumPy and we should broadcast later? Or the current behavior handles this caveat?

NivekT replied:
The current behavior (with the custom padding function) does the bare minimum to increase `x`'s number of dimension if needed. Like NumPy, the broadcasting of two tensors happens **at a later time after** the differences are taken.

I believe if we call something like `x.expand(y.sizes())` or `x.expand(broadcast_tensors(x,y)[0])` here, instead of using the padding function, we would be broadcasting **prior** to the differences being taken and therefore it will diverge from NumPy.

mruberry replied:
That could be. Let's just clearly document this concern / rationale then so people reading this code in the future (like us 1-2 years from now) will remember why we did it

NivekT replied:
The current one just got merged. I am going to add another commit to make sure the comment is there.

mruberry reviewed (approved):
Hey @NivekT! Overall looks good and thanks for the fix. Final question: can expand be used instead of adding a new `add_padding_to_shape` function?
